RecoLocalCalo/HGCalRecProducers/plugins/MergeClusterProducer.cc

```diff
@@ -0,0 +1,207 @@
+  EEclusters_token_ = consumes<std::vector<reco::CaloCluster>>(ps.getParameter<edm::InputTag>("layerClustersEE"));
+  HSiclusters_token_ = consumes<std::vector<reco::CaloCluster>>(ps.getParameter<edm::InputTag>("layerClustersHSi"));
+  HSciclusters_token_ = consumes<std::vector<reco::CaloCluster>>(ps.getParameter<edm::InputTag>("layerClustersHSci"));
+
+  produces<std::vector<float>>("InitialLayerClustersMask");
+  produces<std::vector<reco::BasicCluster>>();
+  produces<std::vector<reco::BasicCluster>>("sharing");
+  //time for layer clusters
+  produces<edm::ValueMap<std::pair<float, float>>>(timeClname_);
+}
+
+void MergeClusterProducer::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
+  // hgcalMergeLayerClusters
+  edm::ParameterSetDescription desc;
+  //layer clusters
+  desc.add<edm::InputTag>("layerClustersEE", edm::InputTag("hgcalLayerClustersEE"));
+  desc.add<edm::InputTag>("layerClustersHSi", edm::InputTag("hgcalLayerClustersHSi"));
+  desc.add<edm::InputTag>("layerClustersHSci", edm::InputTag("hgcalLayerClustersHSci"));
+
+  //time
+  desc.add<edm::InputTag>("time_layerclustersEE", edm::InputTag("hgcalLayerClustersEE", "timeLayerCluster"));
+  desc.add<edm::InputTag>("time_layerclustersHSi", edm::InputTag("hgcalLayerClustersHSi", "timeLayerCluster"));
+  desc.add<edm::InputTag>("time_layerclustersHSci", edm::InputTag("hgcalLayerClustersHSci", "timeLayerCluster"));
+
+  desc.add<std::string>("timeClname", "timeLayerCluster");
+  descriptions.add("hgcalMergeLayerClusters", desc);
+}
+
+void MergeClusterProducer::produce(edm::Event &evt, const edm::EventSetup &es) {
+  //merge clusters
+  std::unique_ptr<std::vector<reco::BasicCluster>> clusters(new std::vector<reco::BasicCluster>);
+  createMerge(evt, EEclusters_token_, HSiclusters_token_, HSciclusters_token_, *clusters);
+  //put new clusters to event
+  auto clusterHandle = evt.put(std::move(clusters));
+
+  //create layer cluster mask
+  std::unique_ptr<std::vector<float>> layerClustersMask(new std::vector<float>);
+  layerClustersMask->resize(clusterHandle->size(), 1.0);
+  //put it into event
+  evt.put(std::move(layerClustersMask), "InitialLayerClustersMask");
+
+  //time
+  std::vector<std::pair<float, float>> times;
+  mergeTime(evt, clusterHandle->size(), times);
+
+  auto timeCl = std::make_unique<edm::ValueMap<std::pair<float, float>>>();
+  edm::ValueMap<std::pair<float, float>>::Filler filler(*timeCl);
+  filler.insert(clusterHandle, times.begin(), times.end());
+  filler.fill();
+  evt.put(std::move(timeCl), timeClname_);
+}
+void MergeClusterProducer::mergeTogether(std::vector<reco::CaloCluster> &merge,
+                                         const std::vector<reco::CaloCluster> &EE,
+                                         const std::vector<reco::CaloCluster> &HSi,
+                                         const std::vector<reco::CaloCluster> &HSci) {
+  auto clusterSize = EE.size() + HSi.size() + HSci.size();
+  merge.reserve(clusterSize);
```

Comment:
I guess this should be
```
merge.reserve(merge.size() + clusterSize);
```
(the function does not guarantee that the input `merge` is empty)

Reply:
(yes, pending, I didn't hold the HLT signature because of this since it concerns RECO code)

Reply:
It is not pending since the `merge`, while not enforced to be empty, it is always empty. I do not think this is a problem. Rather than the proposed solution, if you are really uncomfortable with the code as it is, I'd put an assert on the "emptiness" of the incoming `merge` collection.

Reply:
No opinion, okay with me. "Pending" simply meant that we were looking for a clarification. Thanks for adding it.